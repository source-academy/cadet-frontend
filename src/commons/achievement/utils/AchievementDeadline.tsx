import { Icon } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';

import { DeadlineColors } from '../../../features/achievement/AchievementConstants';
import { AchievementAbility } from '../../../features/achievement/AchievementTypes';
import { isExpired, prettifyDeadline, timeFromExpired } from './DateHelper';

type AchievementDeadlineProps = {
  deadline?: Date;
  ability: AchievementAbility;
};

const twoDays = new Date(0, 0, 2).getTime() - new Date(0, 0, 0).getTime();

function AchievementDeadline(props: AchievementDeadlineProps) {
  const { deadline, ability } = props;

  // red deadline color for core achievements that are expiring in less than 2 days
  const deadlineColor =
    ability === AchievementAbility.CORE &&
    deadline !== undefined &&
    !isExpired(deadline) &&
    timeFromExpired(deadline) <= twoDays
      ? DeadlineColors.RED
      : DeadlineColors.BLACK;

  return (
    <div className="deadline">
      <Icon icon={IconNames.STOPWATCH} color={deadlineColor} />
      <span style={{ color: deadlineColor }}>
        <p>{prettifyDeadline(deadline)}</p>
      </span>
    </div>
  );
}

export default AchievementDeadline;
