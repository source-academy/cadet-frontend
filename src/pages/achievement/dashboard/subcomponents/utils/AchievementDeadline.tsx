import { Icon } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import React from 'react';

import { prettifyDeadline } from './DateHelper';

type AchievementDeadlineProps = {
  deadline?: Date;
};

function AchievementDeadline(props: AchievementDeadlineProps) {
  const { deadline } = props;

  return (
    <div className="deadline">
      <Icon icon={IconNames.STOPWATCH} />
      <p>{prettifyDeadline(deadline)}</p>
    </div>
  );
}

export default AchievementDeadline;
