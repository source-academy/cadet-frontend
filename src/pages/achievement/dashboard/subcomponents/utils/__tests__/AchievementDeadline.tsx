import { mount } from 'enzyme';
import React from 'react';

import AchievementDeadline from '../AchievementDeadline';

const undefinedDeadlineProps = {
  deadline: undefined
};

const expiredDeadlineProps = {
  deadline: new Date()
};

const urgentDeadlineProps = {
  deadline: new Date(Date.now() + 1000 * 60 * 60 * 2)
};

const daysDeadlineProps = {
  deadline: new Date(Date.now() + 1000 * 60 * 60 * 25)
};

const weeksDeadlineProps = {
  deadline: new Date(Date.now() + 1000 * 60 * 60 * 200)
};

test('AchievementDeadline component renders correctly', () => {
  const expiredDeadline = <AchievementDeadline {...expiredDeadlineProps} />;
  const tree_1 = mount(expiredDeadline);
  expect(tree_1.debug()).toMatchSnapshot();

  const urgentDeadline = <AchievementDeadline {...urgentDeadlineProps} />;
  const tree_2 = mount(urgentDeadline);
  expect(tree_2.debug()).toMatchSnapshot();

  const daysDeadline = <AchievementDeadline {...daysDeadlineProps} />;
  const tree_3 = mount(daysDeadline);
  expect(tree_3.debug()).toMatchSnapshot();

  const weeksDeadline = <AchievementDeadline {...weeksDeadlineProps} />;
  const tree_4 = mount(weeksDeadline);
  expect(tree_4.debug()).toMatchSnapshot();

  const undefinedDeadline = <AchievementDeadline {...undefinedDeadlineProps} />;
  const tree = mount(undefinedDeadline);
  expect(tree.debug()).toMatchSnapshot();
});
