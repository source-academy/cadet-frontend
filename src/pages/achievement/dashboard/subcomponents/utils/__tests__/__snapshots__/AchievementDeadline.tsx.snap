// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AchievementDeadline component renders correctly 1`] = `
"<AchievementDeadline deadline={{...}}>
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill={[undefined]} data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <p>
      Expired
    </p>
  </div>
</AchievementDeadline>"
`;

exports[`AchievementDeadline component renders correctly 2`] = `
"<AchievementDeadline deadline={{...}}>
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill={[undefined]} data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <p>
      2 Hours
    </p>
  </div>
</AchievementDeadline>"
`;

exports[`AchievementDeadline component renders correctly 3`] = `
"<AchievementDeadline deadline={{...}}>
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill={[undefined]} data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <p>
      2 Days
    </p>
  </div>
</AchievementDeadline>"
`;

exports[`AchievementDeadline component renders correctly 4`] = `
"<AchievementDeadline deadline={{...}}>
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill={[undefined]} data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <p>
      2 Weeks
    </p>
  </div>
</AchievementDeadline>"
`;

exports[`AchievementDeadline component renders correctly 5`] = `
"<AchievementDeadline deadline={[undefined]}>
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill={[undefined]} data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <p>
      Unlimited
    </p>
  </div>
</AchievementDeadline>"
`;
